Intermittent NRE when using CallTarget

Public copy of internal bug [451463](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/451463).

Builds have failed with this callstack ~20 times in the last 7 or so months in the internal VS build. There's no known simplification, so it's pretty hard to track down.

```
F:\dd\src\ConfigData\BuildTargets\Microsoft.DevDiv.targets(1601,5): error MSB4018: The "CallTarget" task failed unexpectedly. F:\dd\src\ConfigData\BuildTargets\Microsoft.DevDiv.targets(1601,5): error MSB4018: System.NullReferenceException: Object reference not set to an instance of an object.
F:\dd\src\ConfigData\BuildTargets\Microsoft.DevDiv.targets(1601,5): error MSB4018: at Microsoft.Build.Execution.ProjectItemInstance.TaskItem.SetMetadataObject(String name, String metadataValueEscaped, Boolean allowItemSpecModifiers) 
F:\dd\src\ConfigData\BuildTargets\Microsoft.DevDiv.targets(1601,5): error MSB4018: at Microsoft.Build.Execution.ProjectItemInstance.TaskItem.SetMetadata(String metadataName, String metadataValueEscaped) F:\dd\src\ConfigData\BuildTargets\Microsoft.DevDiv.targets(1601,5): error MSB4018: at Microsoft.Build.BackEnd.MSBuild.d__77.MoveNext() 
F:\dd\src\ConfigData\BuildTargets\Microsoft.DevDiv.targets(1601,5): error MSB4018: --- End of stack trace from previous location where exception was thrown --- 
F:\dd\src\ConfigData\BuildTargets\Microsoft.DevDiv.targets(1601,5): error MSB4018: at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) 
F:\dd\src\ConfigData\BuildTargets\Microsoft.DevDiv.targets(1601,5): error MSB4018: at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) 
F:\dd\src\ConfigData\BuildTargets\Microsoft.DevDiv.targets(1601,5): error MSB4018: at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task) 
F:\dd\src\ConfigData\BuildTargets\Microsoft.DevDiv.targets(1601,5): error MSB4018: at Microsoft.Build.BackEnd.TaskBuilder.d__26.MoveNext()
```

@jeffkl and I looked at the [code that's throwing](https://github.com/Microsoft/msbuild/blob/dd134d0a917b870b6a6e805984909a8c341a8083/src/Build/Instance/ProjectItemInstance.cs#L1680-L1686) on Monday. It guards against an NRE:

```c#
_directMetadata = _directMetadata ?? new CopyOnWritePropertyDictionary<ProjectMetadataInstance>();
ProjectMetadataInstance metadatum = new ProjectMetadataInstance(name, metadataValueEscaped, allowItemSpecModifiers /* may not be built-in metadata name */);
_directMetadata.Set(metadatum);
```

So it seems like a race condition, where something is [resetting](https://github.com/Microsoft/msbuild/blob/dd134d0a917b870b6a6e805984909a8c341a8083/src/Build/Instance/ProjectItemInstance.cs#L1306) metadata on an item concurrently with calling `SetMetadataObject`. That's surprising, since this code is fairly single-threaded. There are some `async` methods involved around the [legacy `CallTarget`](https://github.com/Microsoft/msbuild/blob/497d797007410f43e1d7501b7d8254313b07b2a7/src/Build/BackEnd/Components/RequestBuilder/TaskHost.cs#L844-L859) handling, which seems to be involved in every case.

--- a/src/Build/Collections/CopyOnWritePropertyDictionary.cs
+++ b/src/Build/Collections/CopyOnWritePropertyDictionary.cs
@@ -456,11 +456,24 @@ namespace Microsoft.Build.Collections
         /// </summary>
         internal bool Remove(string name)
         {
+            return Remove(name, clearIfEmpty: false);
+        }
+
+        /// <summary>
+        /// Removes any property with the specified name.
+        /// Returns true if the property was in the collection, otherwise false.
+        /// </summary>
+        internal bool Remove(string name, bool clearIfEmpty)
+        {
             ErrorUtilities.VerifyThrowArgumentLength(name, "name");
 
             lock (_properties)
             {
                 bool result = _properties.Remove(name);
+                if (clearIfEmpty && _properties.Count == 0)
+                {
+                    _properties.Clear();
+                }
                 return result;
             }
         }

--- a/src/Build/Instance/ProjectItemInstance.cs
+++ b/src/Build/Instance/ProjectItemInstance.cs
@@ -1235,7 +1235,6 @@ namespace Microsoft.Build.Execution
 
                 if (_directMetadata != null)
                 {
-                    Debug.Assert(_directMetadata.Count != 0, "We should not waste a dictionary for no metadata");
                     metadatum = _directMetadata[metadataName];
                     if (metadatum != null)
                     {
@@ -1345,8 +1344,7 @@ namespace Microsoft.Build.Execution
                     ProjectInstance.VerifyThrowNotImmutable(destinationAsTaskItem._isImmutable);
 
                     // This optimized path is hit most often
-                    Debug.Assert(_directMetadata == null || _directMetadata.Count != 0, "We should not waste a dictionary for no metadata");
-                    destinationAsTaskItem._directMetadata = (_directMetadata == null) ? null : _directMetadata.DeepClone(); // copy on write!
+                    destinationAsTaskItem._directMetadata = _directMetadata?.DeepClone(); // copy on write!
 
                     // If the destination item already has item definitions then we want to maintain them
                     // But ours will be of less precedence than those already on the item

--- a/src/Build/Instance/ProjectItemInstance.cs
+++ b/src/Build/Instance/ProjectItemInstance.cs
@@ -1296,16 +1296,8 @@ namespace Microsoft.Build.Execution
             {
                 ProjectInstance.VerifyThrowNotImmutable(_isImmutable);
 
-                if (_directMetadata != null)
-                {
-                    _directMetadata.Remove(metadataName);
-
-                    // If the metadata was all removed, toss the dictionary
-                    if (_directMetadata.Count == 0)
-                    {
-                        _directMetadata = null;
-                    }
-                }
+                // If the metadata was all removed, toss the dictionary
+                _directMetadata?.Remove(metadataName, clearIfEmpty: true);
             }
 
             /// <summary>

